CRM/Fieldmetadata/Normalizer.php

```diff
@@ -79,6 +80,92 @@ function setVisibilityForFields(&$fields) {
     }
   }
 
+  /**
+   * This will go through and add minDate and maxDate keys
+   * to each date field found. It will work for both custom
+   * fields and core fields.
+   *
+   * @param array $fields
+   * @return void
+   */
+  function fetchAndNormalizeDates(&$fields) {
+   
+    static $contactTypes = [];
+    static $dates = [];
+    static $formats = [];
+
+    $dateWidgets = array('Date', 'DateTime', 'Select Date', 'crm-ui-datepicker');
+    $thisYear = date('Y');
+
+    foreach ($fields as &$field) {
+      // skip non-date fields
+      if (!in_array($field['widget'], $dateWidgets)) {
+        continue;
+      }
+
+      // only do the lookup once
+      if (empty($dates[$field['name']])) {
+
+        // custom fields are little bit easier
+        if (strpos($field['name'], 'custom_') === 0) {
+          $customField = civicrm_api3('CustomField', 'getsingle', array(
+            'id' => substr($field['name'], 7),
+            'return' => 'start_date_years, end_date_years',
+          ));
+          $start = $customField['start_date_years'];
+          $end = $customField['end_date_years'];
+        }
+        // core fields
+        else {
+          // get contact types, i.e. potential entity name
+          if (empty($contactTypes)) {
+            $api = civicrm_api3('ContactType', 'get', array(
+              'sequential' => 1,
+              'is_active' => 1,
+            ));
+            foreach ($api['values'] as $contactType) {
+              $contactTypes[] = $contactType['name'];
+            }
+          }
+          
+          // look up the field first
+          $entity = in_array($field['entity'], $contactTypes) ? 'Contact' : $field['entity'];
+          $api = civicrm_api3($entity, 'getfield', array(
+            'name' => $field['name'],
+            'action' => 'get',
+          ));
+          
+          if (!$api['is_error'] && !empty($api['values']['html']['formatType'])) {
```

Comment:
Isn't this check unecessary since we are using civicrm_api3 and if there is an error, we won't reach this line because of the exception? Should we be using civicrm_api() instead and keep this check so that we can continue in the case of an error?